### DISCUSSION TITLE
How to wait for page reload?

### DISCUSSION THREAD

**Discussion by enumag** [php-webdriver/php-webdriver#821] [Q&A] [answered]
I use `executeScript` to fill and submit a form. Then I need to wait for the page to reload because I need the URL I'm redirected to. But I can't find any method to wait until then. Can you give me an advice?

**Comment by OndraM** [comment]
Hi, you have several options, depending on behavior of the form you submit.

But the way to go is using explicit wait. For example you can use `refreshed()` condition, [see docs.](https://github.com/php-webdriver/php-webdriver/wiki/HowTo-Wait#special-conditions)

```php
$driver->wait()->until(WebDriverExpectedCondition::refreshed());
```

In case this doesn't works for your scenario, you can wait for URL change (`urlContains()` etc.) or appearance of eg. "Form submitted" message or something similar (`elementTextContains()`).

**Comment by enumag** [comment]
`WebDriverExpectedCondition::refreshed()` requires some argument. What should I pass there?

**Comment by enumag** [comment]
Using URL or text is something I want to avoid. Those things can change at any time on the website so I don't want to rely on that. I just want to wait until the page is reloaded after form submit.

**Comment by OndraM** [comment, marked as answer]
Oh, sorry, I wrote the example from memory. It requires some other condition, which is expected to throw "stale element" condition (meaning the original element instance is no longer present on the page). Yes, this means you need *some* element which is present on the page and will be redrawn after form is submitted.

For example:

```
$driver->wait()->until(
    WebDriverExpectedCondition::refreshed(
        WebDriverExpectedCondition::presenceOfElementLocated(
            WebDriverBy::cssSelector('#formHeader')
        )
    )
);
````

You may also try using `executeScript()` and wait for some browser events (page ready etc.), but I'd say its overcomplication.


> Using URL or text is something I want to avoid. Those things can change at any time on the website so I don't want to rely on that. I just want to wait until the page is reloaded after form submit.

Sure, but also the element selector can change "any time". So you must choose the most stable way depending on your needs.

**Comment by enumag** [comment]
> Sure, but also the element selector can change "any time". So you must choose the most stable way depending on your needs.

That's what I'm talking about. I don't want ANY element, text, url or whatever. I only want to wait for the refresh, I don't care at all what is on the new page... depending on any specific condition is useless overcomplication and prone to break in the future.

**Comment by OndraM** [comment]
So how do you then want the webdriver to know what are you waiting for to be "refreshed"? 

> depending on any specific condition is useless overcomplication

I don't agree  - depending on *specific* *explicit* wait conditions (which are based on the exact behavior of the page) is actually the best attitude for stable functional tests. Sure, you must balance "possibility of change (stability) of the condition"  with "too generic false-positive-prone condition".

Also, if you whole page contents reload, you can wait for html, body or form element to be reloaded - I don't see this overcomplicated in any way.

Other possibility if the behavior is that you are redirected to some other URL after form submit is to make custom condition and wait for the URL change like this (not tested):

```
// CustomCondition.php

... 
public static function urlIsNot($currentUrl)
{
    return new WebDriverExpectedCondition(
        function (WebDriver $driver) use ($currentUrl) {
            return $currentUrl !== $driver->getCurrentURL();
        }
    );
}

// usage in tests
$driver->wait()->until(CustomCondition::urlIsNot($driver->getCurrentURL());
```

**Comment by enumag** [comment]
> So how do you then want the webdriver to know what are you waiting for to be "refreshed"?

I'm not waiting for anything specific, just that the form is submitted and I'm on the "new page" after form submission...

> Also, if you whole page contents reload, you can wait for html, body or form element to be reloaded - I don't see this overcomplicated in any way.

Okay, how do I wait for `html` element to be reloaded then? :thinking: I mean there is an html element on the original page as well...

Sure there is likely some element that is present on one and isn't on the other or vice-versa but those elements are not actually important to me so I don't want my script to break if their ID changes or something.

> Other possibility if the behavior is that you are redirected to some other URL after form submit is to make custom condition and wait for the URL change like this (not tested):

Sadly the URL doesn't always change so I can't rely on that.

----

For more context, I'm trying to switch from https://github.com/chrome-php/headless-chromium-php to Selenium. In that library it was quite simple:

```
		$page->evaluate($javascript)
			->waitForResponse($timeout)
			->waitForPageReload(Page::LOAD, $timeout);
```

So I tried to check how they implement it. It seems to be more difficult than I expected: https://github.com/chrome-php/headless-chromium-php/blob/master/src/Page.php#L325-L375

I'm guessing Selenium doesn't have anything similar?

**Comment by enumag** [comment]
Okay. It seems I managed to find an ugly workaround...:

```php
$driver->executeScript('window.waiting = true');
$driver->executeScript($fillAndSubmitForm);
$driver->wait()->until(fn (RemoteWebDriver $driver) => $driver->executeScript('return window.waiting === undefined'));
```

Do you have some ideas how to improve this? Ideally without JS.

**Comment by OndraM** [comment]
To wait until instance of previous body element is missing (which basically means new page was loaded) try this:

```php
$bodyElement = $driver->findElement(WebDriverBy::cssSelector('body')); // here you get instance of *current* body element

$driver->executeScript('...'); // here you cause the page to by submitted

$driver->wait()->until(
    WebDriverExpectedCondition::stalenessOf($bodyElement) // this condition wait until instance of the $bodyElement is "stale" ie. missing on the page
);
```

**Comment by enumag** [comment]
Thanks! I still consider this an ugly workaround (would prefer something that doesn't require a separate code before `executeScript`) but it works.

**Comment by OndraM** [comment]
Glad to help.

> would prefer something that doesn't require a separate code before `executeScript`

Well, this may be possible, I mentioned few other options, but it really depends on the exact site and scenario, they are not applicable in all situations.

**Comment by andrewnicols** [comment]
The issue here is that you are submitting the form using `executeScript` rather than a user interaction.

The WebDriver specification returns immediately upon completing the `executeScript()` invocation. If you were using a user interaction such as clicking on a button then the `click()` invocation would only return once the page has changed (though likely before the page is ready if you have any heavy setup).
